### ISSUE TITLE
Optional types doesn't work with custom actions

### ISSUE THREAD

**Issue by stasbel** [omni-us/jsonargparse#45]
This does work:

```python
class CustomAction(argparse.Action):
    def __call__(self, parser, args, values, option_string=None):
        print('Custom action!')
        setattr(args, self.dest, values)

parser = jsonargparse.ArgumentParser()
parser.add_argument('--path', type=str, action=CustomAction)

parser.parse_args(['--path=~/data'])   
```

But this doesn't:

```python3
class CustomAction(argparse.Action):
    def __call__(self, parser, args, values, option_string=None):
        print('Custom action!')
        setattr(args, self.dest, values)

parser = jsonargparse.ArgumentParser()
parser.add_argument('--path', type=Optional[str], action=CustomAction)

parser.parse_args(['--path=~/data'])   
```

**Comment by mauvilsa** [comment]
The support for type hints in jsonargparse is implemented by converting the type to a json schema via the ActionJsonSchema. Therefore it is not supposed to be supported to have a type hint and some custom action. I will add a check so that an error is raised if a type hint is used with a custom action.

Looking at the example it seems you want an optional path. This is supported with the path types, e.g.

```python
import jsonargparse
from jsonargparse.typing import Path_fc
from typing import Optional

parser = jsonargparse.ArgumentParser()
parser.add_argument('--path', type=Optional[Path_fc])

parser.parse_args(['--path=data'])
```

The `Path_fc` would be for a file that can be created. There are more predefined path types or you can create new ones depending on whether you need file/directory/read/write/execute, etc. The home prefix `~/` is not currently supported. I will take a note to add this later.

**Comment by stasbel** [comment]
that's nice, but what if I want to add custom processing logic along with supporting Optional typing?

**Comment by mauvilsa** [comment]
The type hint support is designed to be able to handle arbitrarily complex types. Depending on what you want, being able to add custom processing logic might require fiddling with the internal working of this. Not good because if something is changed in jsonargparse it would very likely break your custom processing. Adding support for this custom processing logic would require a well thought design and clear use cases. It would be great if you could explain exactly what kind of logic you want to add.

Anyway, in general the idea is that whenever possible, logic should be reduced to implementing a type class. That is, you implement a new type, say `MyType`, and then you register this type for use in the parsers. Preferably `str(MyType('value')) == 'value'` should be true. Alternatively when registering you can give serializer and/or deserializer functions. The good thing about implementing types is that they can be used independently of the argument parser, for example to use as type hints in function parameters. And also you can use it inside arbitrary type hints, so not just `Optional[MyType]` but possible things like `Optional[Dict[str, List[MyType]]]` or whatever.

**Comment by mauvilsa** [comment]
With the latest commit in master the Path class now supports the home prefix `~`.

**Comment by mauvilsa** [comment]
I have just released v3.8.0 which includes the home prefix support.
